pingora-core/src/protocols/tls/s2n/client.rs

```diff
@@ -0,0 +1,44 @@
+// Copyright 2025 Cloudflare, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! S2N client specific implementation
+
+use crate::protocols::tls::{AutoFlushableStream, S2NConnectionBuilder, TlsStream};
+use crate::protocols::IO;
+use pingora_error::ErrorType::TLSHandshakeFailure;
+use pingora_error::{Error, Result};
+use pingora_s2n::TlsConnector;
+
+// Perform the TLS handshake for the given connection with the given configuration
+pub async fn handshake<S: IO>(
+    connector: &TlsConnector<S2NConnectionBuilder>,
+    domain: &str,
+    stream: S,
+) -> Result<TlsStream<S>> {
+    // Wrap incoming stream in an auto flushable stream with auto flush enabled because
+    // s2n-tls doesn't invoke flush after writing to the connection. This would result in
+    // the handshake hanging and timing on streams with write buffering.
+    let auto_flushable_stream = AutoFlushableStream::new(stream, true);
+    let mut s2n_stream = connector
+        .connect(domain, auto_flushable_stream)
+        .await
+        .map_err(|e| {
+            let context = format!("TLS connect() failed: {e}, SNI: {domain}");
+            Error::explain(TLSHandshakeFailure, context)
+        })?;
+
+    // Disable auto-flush to not interfere with write buffering going forward.
```

Comment:
This is a good catch. How did you find these behaviors? hung connection troubleshooting or did the s2n docs give you a call out?

Reply:
LOTS of troubleshooting on a hung connection. I initially thought that it was a problem with the customized async Runtime in Pingora and the waker system not properly waking the executor to drive the underlying Futures. I only figured it out after digging into the underlying stream implementation (`pingora_core::protocols::l4::stream::Stream`) and seeing that it was buffered before I figured out what was actually happening.

I plan on following up on this with the `s2n-tls` folks. Since we don't have any way to influence the handshake process here, I had to resort to wrapping the connection with the Auto Flushable stream. Ideally `s2n-tls` would handle flushing instead of assuming an un-buffered stream and we can remove this logic in the future.

Reply:
haha, i bet that fell damn good once you found the solve.